chips/sam4l/src/adc.rs

```diff
@@ -1,209 +1,766 @@
+            // only one sample at a time
+            self.return_buffers(Some(buffer1), Some(buffer2), ReturnCode::EBUSY)
+
+        } else if frequency <= (self.adc_clk_freq.get() / (0xFFFF + 1)) || frequency > 250000 {
+            // can't sample faster than the max sampling frequency or slower than the timer can be
+            // set to
+            self.return_buffers(Some(buffer1), Some(buffer2), ReturnCode::EINVAL)
 
-    fn compute_interval(&self, interval: u32) -> u32 {
-        interval
+        } else if length1 == 0 {
+            // at least need a valid lenght for the for the first buffer full of samples.
+            // Otherwise, what are we doing here?
+            self.return_buffers(Some(buffer1), Some(buffer2), ReturnCode::EINVAL)
+
+        } else {
+            self.active.set(true);
+            self.continuous.set(true);
+
+            // store the second buffer for later use
+            self.next_dma_buffer.replace(buffer2);
+            self.next_dma_length.set(length2);
+
+            // adc sequencer configuration
+            let cfg = (0x7 << 20) | // MUXNEG: ground pad
+                      (channel.chan_num << 16) | // MUXPOS: selection
+                      (0x1 << 15) | // INTERNAL: internal neg
+                      (channel.internal << 14) | // INTERNAL: pos selection
+                      (0x0 << 12) | // RES: 12-bit resolution
+                      (0x1 <<  8) | // TRGSEL: internal timer trigger
+                      (0x0 <<  7) | // GCOMP: no gain compensation
+                      (0x7 <<  4) | // GAIN: 0.5x gain
+                      (0x0 <<  2) | // BIPOLAR: unipolar mode
+                      (0x0 <<  0); // HWLA: right justify value
+            regs.seqcfg.set(cfg);
+
+            // stop timer if running
+            regs.cr.set(0x02);
+
+            // set timer, limit to bounds
+            // f(timer) = f(adc) / (counter + 1)
+            let mut counter = (self.adc_clk_freq.get() / frequency) - 1;
+            counter = cmp::max(cmp::min(counter, 0xFFFF), 0);
+            regs.itimer.set(counter);
+
+            // clear any current status
+            regs.scr.set(0x2F);
+
+            // receive up to the buffer's length samples
+            let dma_len = cmp::min(buffer1.len(), length1);
+
+            // change buffer into a [u8]
+            // this is unsafe but acceptable for the following reasons
+            //  * the buffer is aligned based on 16-bit boundary, so the 8-bit
+            //    alignment is fine
+            //  * the DMA is doing checking based on our expected data width to
+            //    make sure we don't go past dma_buf.len()/width
+            //  * we will transmute the array back to a [u16] after the DMA
+            //    transfer is complete
+            let dma_buf_ptr = unsafe { mem::transmute::<*mut u16, *mut u8>(buffer1.as_mut_ptr()) };
+            let dma_buf = unsafe { slice::from_raw_parts_mut(dma_buf_ptr, buffer1.len() * 2) };
```

Comment:
@alevy Still want you to sign off on this method of buffer handling. (or to tell me that it's horribly wrong for some reason I haven't considered, or that there is a better way to do it)